diff --git c/.github/workflows/rust.yml w/.github/workflows/rust.yml
index 0db8ece..1c168f7 100644
--- c/.github/workflows/rust.yml
+++ w/.github/workflows/rust.yml
@@ -21,7 +21,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest, windows-latest, macos-latest]
-        rust: ['stable', '1.85']
+        rust: ['stable', '1.87']
 
     runs-on: ${{ matrix.os }}
 
diff --git c/Cargo.toml w/Cargo.toml
index 2f82c5d..84fc445 100644
--- c/Cargo.toml
+++ w/Cargo.toml
@@ -8,7 +8,7 @@ authors = ["Kevin Mehall <km@kevinmehall.net>"]
 edition = "2021"
 license = "Apache-2.0 OR MIT"
 repository = "https://github.com/kevinmehall/nusb"
-rust-version = "1.85" # keep in sync with .github/workflows/rust.yml
+rust-version = "1.87" # keep in sync with .github/workflows/rust.yml
 
 [dependencies]
 futures-core = "0.3.29"
@@ -28,7 +28,7 @@ linux-raw-sys = { version = ">= 0.11, <= 0.12", features = ["ioctl"] }
 once_cell = "1.18.0"
 
 [target.'cfg(target_os="windows")'.dependencies]
-windows-sys = { version = ">= 0.60, <= 0.61", features = ["Win32_Devices_Usb", "Win32_Devices_DeviceAndDriverInstallation", "Win32_Foundation", "Win32_Devices_Properties", "Win32_Storage_FileSystem", "Win32_Security", "Win32_System_IO", "Win32_System_Registry", "Win32_System_Threading", "Win32_System_Com"] }
+windows-sys = { version = ">= 0.60, <= 0.61", features = ["Win32_Devices_Usb", "Win32_Devices_DeviceAndDriverInstallation", "Win32_Foundation", "Win32_Devices_Properties", "Win32_Storage_FileSystem", "Win32_Security", "Win32_System_IO", "Win32_System_Ioctl", "Win32_System_Registry", "Win32_System_Threading", "Win32_System_Com"] }
 
 [target.'cfg(target_os="macos")'.dependencies]
 core-foundation = "0.10.1"
diff --git c/src/hotplug.rs w/src/hotplug.rs
index 790725e..1f8d788 100644
--- c/src/hotplug.rs
+++ w/src/hotplug.rs
@@ -31,6 +31,9 @@ pub enum HotplugEvent {
     /// A device has been connected.
     Connected(DeviceInfo),
 
+    /// Serial device connected
+    SerialConnected(DeviceInfo),
+
     /// A device has been disconnected.
     Disconnected(DeviceId),
 }
diff --git c/src/lib.rs w/src/lib.rs
index cd11c10..525cdd7 100644
--- c/src/lib.rs
+++ w/src/lib.rs
@@ -376,6 +376,9 @@ pub fn list_buses() -> impl MaybeFuture<Output = Result<impl Iterator<Item = Bus
 ///         HotplugEvent::Connected(d) => {
 ///             devices.insert(d.id(), d);
 ///         }
+///         HotplugEvent::SerialConnected(d) => {
+///             println!("Seria port connected: {d:?}");
+///         }
 ///         HotplugEvent::Disconnected(id) => {
 ///             devices.remove(&id);
 ///         }
diff --git c/src/platform/macos_iokit/hotplug.rs w/src/platform/macos_iokit/hotplug.rs
index d43ab89..ea06ce7 100644
--- c/src/platform/macos_iokit/hotplug.rs
+++ w/src/platform/macos_iokit/hotplug.rs
@@ -6,6 +6,7 @@ use std::{
 };
 
 use core_foundation::{base::TCFType, dictionary::CFDictionary, runloop::CFRunLoopSource};
+use io_kit_sys::serial::keys::kIOSerialBSDServiceValue;
 use io_kit_sys::{
     kIOMasterPortDefault,
     keys::{kIOFirstMatchNotification, kIOTerminatedNotification},
@@ -52,6 +53,7 @@ pub(crate) struct MacHotplugWatch {
     waker_id: SlabWaker,
     terminated_iter: IoServiceIterator,
     matched_iter: IoServiceIterator,
+    cdc_matched_iter: IoServiceIterator,
     _registration: EventRegistration,
     _notification_port: NotificationPort,
 }
@@ -84,6 +86,14 @@ impl MacHotplugWatch {
             CFDictionary::wrap_under_create_rule(d)
         };
 
+        let cdc_dictionary = unsafe {
+            let d = IOServiceMatching(kIOSerialBSDServiceValue);
+            if d.is_null() {
+                return Err(Error::new(ErrorKind::Other, "IOServiceMatching failed"));
+            }
+            CFDictionary::wrap_under_create_rule(d)
+        };
+
         let notification_port = NotificationPort::new();
         let terminated_iter = register_notification(
             &notification_port,
@@ -97,6 +107,12 @@ impl MacHotplugWatch {
             &waker_id,
             kIOFirstMatchNotification,
         )?;
+        let cdc_matched_iter = register_notification(
+            &notification_port,
+            &cdc_dictionary,
+            &waker_id,
+            kIOFirstMatchNotification,
+        )?;
 
         let source = unsafe {
             CFRunLoopSource::wrap_under_get_rule(IONotificationPortGetRunLoopSource(
@@ -109,6 +125,7 @@ impl MacHotplugWatch {
             waker_id,
             terminated_iter,
             matched_iter,
+            cdc_matched_iter,
             _registration: registration,
             _notification_port: notification_port,
         })
@@ -125,6 +142,14 @@ impl MacHotplugWatch {
             }
         }
 
+        while let Some(s) = self.cdc_matched_iter.next() {
+            if let Some(dev) = probe_device(s) {
+                return Poll::Ready(HotplugEvent::SerialConnected(dev));
+            } else {
+                debug!("failed to probe serial connected device");
+            }
+        }
+
         if let Some(s) = self.terminated_iter.next() {
             if let Some(registry_id) = get_registry_id(&s) {
                 debug!("device {registry_id} disconnected");
diff --git c/src/platform/windows_winusb/enumeration.rs w/src/platform/windows_winusb/enumeration.rs
index 9dc23d5..e6789bc 100644
--- c/src/platform/windows_winusb/enumeration.rs
+++ w/src/platform/windows_winusb/enumeration.rs
@@ -16,6 +16,7 @@ use crate::{
         DESCRIPTOR_TYPE_CONFIGURATION, DESCRIPTOR_TYPE_STRING,
     },
     maybe_future::{blocking::Blocking, MaybeFuture},
+    platform::windows_winusb::util::WCStr,
     BusInfo, DeviceInfo, Error, ErrorKind, InterfaceInfo, UsbControllerType,
 };
 
@@ -52,6 +53,41 @@ pub fn list_buses() -> impl MaybeFuture<Output = Result<impl Iterator<Item = Bus
     })
 }
 
+pub fn probe_serial_device(devinst: DevInst) -> Option<DeviceInfo> {
+    let instance_id = devinst.get_property::<OsString>(DEVPKEY_Device_InstanceId)?;
+    let mut devinst = devinst;
+
+    if instance_id.to_string_lossy().starts_with("USB\\ROOT_HUB") {
+        return None;
+    }
+
+    log::debug!("Probing serial device {}", instance_id.display());
+
+    // Limit the number of iterations to avoid infinite loops,
+    // even though it's unlikely to happen.
+    let n = 100;
+
+    for _ in 0..n {
+        let interfaces = devinst.interfaces(GUID_DEVINTERFACE_USB_DEVICE);
+
+        let paths = interfaces.iter().count();
+
+        if paths > 0 {
+            return probe_device(devinst);
+        }
+
+        let instance_id = devinst.get_property::<OsString>(DEVPKEY_Device_Parent)?;
+
+        let parent_id: WCString = WCString::from(instance_id.as_os_str());
+
+        devinst = DevInst::from_instance_id(&parent_id)?;
+    }
+
+    // The code will go up the device tree until it ends up at the root of the device tree,
+    // where there is no parent. So this part should never be reached.
+    unreachable!("Failed to find parent USB device, and did not reach root of the device tree")
+}
+
 pub fn probe_device(devinst: DevInst) -> Option<DeviceInfo> {
     let instance_id = devinst.get_property::<OsString>(DEVPKEY_Device_InstanceId)?;
     if instance_id.to_string_lossy().starts_with("USB\\ROOT_HUB") {
@@ -279,13 +315,17 @@ pub(crate) fn get_usbccgp_winusb_device_path(child: DevInst) -> Result<WCString,
             Ok(s) => s,
             Err(f) => {
                 if e.kind() == f.kind() {
-                    debug!("Failed to get DeviceInterfaceGUID or DeviceInterfaceGUIDs from registry: {e}");
+                    debug!(
+                        "Failed to get DeviceInterfaceGUID or DeviceInterfaceGUIDs from registry: {e}"
+                    );
                 } else {
-                    debug!("Failed to get DeviceInterfaceGUID or DeviceInterfaceGUIDs from registry: {e}, {f}");
+                    debug!(
+                        "Failed to get DeviceInterfaceGUID or DeviceInterfaceGUIDs from registry: {e}, {f}"
+                    );
                 }
                 return Err(Error::new(
                     ErrorKind::Unsupported,
-                    "Could not find DeviceInterfaceGUIDs in registry. WinUSB driver may not be correctly installed for this interface."
+                    "Could not find DeviceInterfaceGUIDs in registry. WinUSB driver may not be correctly installed for this interface.",
                 ));
             }
         },
diff --git c/src/platform/windows_winusb/hotplug.rs w/src/platform/windows_winusb/hotplug.rs
index 2bcc26c..48ecb97 100644
--- c/src/platform/windows_winusb/hotplug.rs
+++ w/src/platform/windows_winusb/hotplug.rs
@@ -20,11 +20,14 @@ use windows_sys::Win32::{
         Usb::GUID_DEVINTERFACE_USB_DEVICE,
     },
     Foundation::ERROR_SUCCESS,
+    System::Ioctl::GUID_DEVINTERFACE_COMPORT,
 };
 
 use crate::{
     hotplug::HotplugEvent,
-    platform::windows_winusb::{cfgmgr32::get_device_interface_property, util::WCString},
+    platform::windows_winusb::{
+        cfgmgr32::get_device_interface_property, enumeration::probe_serial_device, util::WCString,
+    },
     DeviceId, Error,
 };
 
@@ -35,6 +38,7 @@ use super::DevInst;
 pub(crate) struct WindowsHotplugWatch {
     inner: *mut HotplugInner,
     registration: HCMNOTIFICATION,
+    registration_com: HCMNOTIFICATION,
 }
 
 struct HotplugInner {
@@ -45,6 +49,7 @@ struct HotplugInner {
 #[derive(Debug)]
 enum Action {
     Connect,
+    SerialConnect,
     Disconnect,
 }
 
@@ -86,9 +91,41 @@ impl WindowsHotplugWatch {
             .log_error());
         }
 
+        let mut registration_com = ptr::null_mut();
+        let filter_com = CM_NOTIFY_FILTER {
+            cbSize: size_of::<CM_NOTIFY_FILTER>() as u32,
+            Flags: 0,
+            FilterType: CM_NOTIFY_FILTER_TYPE_DEVICEINTERFACE,
+            Reserved: 0,
+            u: CM_NOTIFY_FILTER_0 {
+                DeviceInterface: CM_NOTIFY_FILTER_0_0 {
+                    ClassGuid: GUID_DEVINTERFACE_COMPORT,
+                },
+            },
+        };
+
+        let cr_com = unsafe {
+            CM_Register_Notification(
+                &filter_com,
+                inner as *mut c_void,
+                Some(hotplug_callback_serial),
+                &mut registration_com,
+            )
+        };
+
+        if cr_com != CR_SUCCESS {
+            return Err(Error::new_os(
+                crate::ErrorKind::Other,
+                "failed to initialize hotplug notifications for serial",
+                cr,
+            )
+            .log_error());
+        }
+
         Ok(WindowsHotplugWatch {
             inner,
             registration,
+            registration_com,
         })
     }
 
@@ -109,6 +146,11 @@ impl WindowsHotplugWatch {
                     return Poll::Ready(HotplugEvent::Connected(dev));
                 };
             }
+            Some((Action::SerialConnect, devinst)) => {
+                if let Some(dev) = probe_serial_device(devinst) {
+                    return Poll::Ready(HotplugEvent::SerialConnected(dev));
+                };
+            }
             Some((Action::Disconnect, devinst)) => {
                 return Poll::Ready(HotplugEvent::Disconnected(DeviceId(devinst)));
             }
@@ -132,6 +174,7 @@ impl Drop for WindowsHotplugWatch {
             // immediately afterward without races.
             // [1]: https://learn.microsoft.com/en-us/windows/win32/api/cfgmgr32/nf-cfgmgr32-cm_unregister_notification
             CM_Unregister_Notification(self.registration);
+            CM_Unregister_Notification(self.registration_com);
             drop(Box::from_raw(self.inner));
         }
     }
@@ -178,3 +221,42 @@ unsafe extern "system" fn hotplug_callback(
 
     ERROR_SUCCESS
 }
+
+unsafe extern "system" fn hotplug_callback_serial(
+    _hnotify: HCMNOTIFICATION,
+    context: *const ::core::ffi::c_void,
+    action: CM_NOTIFY_ACTION,
+    eventdata: *const CM_NOTIFY_EVENT_DATA,
+    _eventdatasize: u32,
+) -> u32 {
+    let inner = unsafe { &*(context as *const HotplugInner) };
+
+    let action = match action {
+        CM_NOTIFY_ACTION_DEVICEINTERFACEARRIVAL => Action::SerialConnect,
+        _ => {
+            debug!("Hotplug callback: unknown or unhandled action {action}");
+            return ERROR_SUCCESS;
+        }
+    };
+
+    let device_interface =
+        unsafe { WCStr::from_ptr(addr_of!((*eventdata).u.DeviceInterface.SymbolicLink[0])) };
+
+    let device_instance =
+        get_device_interface_property::<WCString>(device_interface, DEVPKEY_Device_InstanceId)
+            .unwrap();
+    let devinst = DevInst::from_instance_id(&device_instance).unwrap();
+
+    if let Some(parent) = devinst.parent() {
+        debug!("Hotplug com callback: action={action:?}, instance={device_instance}");
+        inner.events.lock().unwrap().push_back((action, parent));
+        if let Some(w) = inner.waker.lock().unwrap().take() {
+            w.wake()
+        }
+    } else {
+        debug!(
+            "Hotplug com callback no parent devinst found: action={action:?}, instance={device_instance}"
+        );
+    }
+    ERROR_SUCCESS
+}
